### ISSUE TITLE
[Data] retrieve written paths from `Dataset.write_datasource`

### ISSUE THREAD

**Issue by harelwa** [ray-project/ray#34444]
### What happened + What you expected to happen

**UPDATE**

TL;DR how can retrieve the actual paths of files written in a distributed execution by `Dataset.write_datasource` ( or `write_parquet` for that matter ) ?

**Use case**

for spreading massive writes to S3 (physically) partitioned buckets, I have implemented a custom `BlockWritePathProvider`  paths naming function (a hash function).

I need to get these paths back, as these paths do not share a path `prefix` you can later use to load the dataset. It's loaded only with an explicit list for files.

I have tried to work around this by _storing_ these paths in a dedicated data member of my _custom_ `BlockWritePathProvider` class instance. 

As expected, this works only for local writes (i.e. - when `RAY_DATASET_FORCE_LOCAL_METADATA` env var is set). But irrelevant for distributed writes which is my use case.

How would you suggest to solve this ?

Thanks,
Harel

### Versions / Dependencies

`OS` Linux
`Python` 3.8
`Ray` 2.3.1

### Reproduction script

api usage of `Dataset.write_parquet` reproduces this

### Issue Severity

High: It blocks me from completing my task.

**Comment by ericl** [comment]
In the longer run, I think we should return those paths as output of write (cc @c21 for thoughts).

> As expected, this works only for local writes (i.e. - when RAY_DATASET_FORCE_LOCAL_METADATA env var is set). But irrelevant for distributed writes which is my use case.

One workaround here is to use a named actor to collect the results (see https://docs.ray.io/en/master/ray-core/actors/named-actors.html#get-or-create-a-named-actor). Basically, create the actor beforehand, have your BlockPathWriteProvider send the computed paths to the actor, and then query the actor for the paths after the write completes.

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
